Add vertical mixing coefficients module

katsmith133 opened the pull request:
Add the `VertMix` class to Omega, which consists of:

Methods to calculate the vertical diffusivity and viscosity using (for now just) the `PP` option, which can linearly add contributions from background, convective, and shear mixing based upon user choices in the config file. The `VertMix` options and parameter values are set in the yaml config file. There is a dummy `KPP` option in there now to set up the structure for adding it in the near future (I can remove if we think this is messy). This PR includes tests for each individual mixing process (background, convective, and shear), as well as one that linearly adds all of them together. Because both the convective and shear mixing require the Brunt Vaiasla frequency, this PR also adds the calculation of `BruntVaisalaFreq` to the EOS class and adds corresponding tests. The `BruntVaisalaFreq` is calculated in the way that was in MPAS-O if the Linear EOS is used (i.e. with linear coefficients and the derivative based upon changes in `z`) and is calculated in the same way the TEOS-10 package calculates it if the TEOS-10 EOS is used (i.e. with non-linear coefficients and the derivative based upon changes in `p`).

Ran ctests successfully on pm-cpu, pm-gpu, and chrysalis.

Checklist
* [x] Documentation:
  * [x] User's Guide has been updated
  * [x] Developer's Guide has been updated
  * [x] Documentation has been [built locally](https://e3sm-project.github.io/Omega/omega/develop/devGuide/BuildDocs.html) and changes look as expected
* [x] Building
  * [x] CMake build does not produce any new warnings from changes in this PR
* [x] Testing
  * [x] A comment in the PR documents testing used to verify the changes including any tests that are added/modified/impacted.
  * [x] CTest unit tests for new features have been added per the approved design.
  * [x] Unit tests have passed. Please provide a relevant CDash build entry for verification.

katsmith133 committed:
adds basic vertical mixing class structure

katsmith133 committed:
adds bvf to eos

katsmith133 committed:
adding n2 and vertical mixing

katsmith133 committed:
fixes bvf calc and vertmix calcs

katsmith133 committed:
fixes bvf tests

katsmith133 committed:
fixes out of bounds issues

katsmith133 committed:
fixes vertmix test and adds documentation

katsmith133 committed:
adjusting testing range to nvertlayers

katsmith133 committed:
fixing out of bounds issue

katsmith133 committed:
removes template changes

philipwjones commented on components/omega/configs/Default.yml:
```diff
@@ -59,6 +59,20 @@ Omega:
       DRhoDS: 0.8
       RhoT0S0: 1000.0
+  VertMix:
+    VertMixType: PP
```
Do we need a VertMixType with the individual enable sections? KPP could just have a separate enable for the K-profile and non-local stuff.

katsmith133 replied:
That's a good point... I will adjust to have them all have individual enable sections rather than use the VertMixType route.

philipwjones commented on components/omega/doc/devGuide/VerticalMixingCoeff.md:
```diff
@@ -0,0 +1,96 @@
+
+## VertMix Type
+
+An enumeration listing of the schemes is provided (`KPP` is not yet functional). It needs to be extended every time a `VertMix` type is added. It is used to identify which VertMix method is to be used at run time.
```
See prev comment on whether VertMixType is needed

philipwjones commented on components/omega/doc/userGuide/VerticalMixingCoeff.md:
```diff
@@ -0,0 +1,67 @@
+dependent shear mixing value based upon Pacanowski and Philander (1981). These are describe a bit
+more in detail below. The second vertical mixing option, `KPP`, is not yet functional, but will be added soon.
+
+The user-configurable options are: `VertMixType` (choose either `PP` or `KPP`), as well as the following parameters in the yaml configuration file for each mixing process within the `PP` option:
```
Same comment on VertMixType

philipwjones commented on components/omega/src/ocn/Eos.cpp:
```diff
@@ -218,6 +270,18 @@ void Eos::defineFields() {
+                     "Brunt-Vaisala frequency squared",           // Long Name
+                     "s-2",                                       // Units
+                     "sea_water_brunt_vaisala_frequency_squared", // CF-ish Name
+                     0.0,       // Min valid value
```
BVSq can be negative so min val should be < 0

katsmith133 replied:
Good catch, thanks! I will adjust this.

philipwjones commented on components/omega/src/ocn/VertMix.cpp:
```diff
@@ -0,0 +1,297 @@
+
+   /// Dispatch to the correct VertMix calculation
+   if (VertMixChoice == VertMixType::PP) {
+      LOG_INFO("VertMix:: into PP computeVertMix");
```
These look like debug/test statements? We don't want to print stuff every time...

philipwjones commented on components/omega/src/ocn/VertMix.h:
```diff
@@ -0,0 +1,210 @@
+            }
+            ShearSquared =
+                ShearSquared /
+                Kokkos::pow(ZMid(ICell, K - 1) - ZMid(ICell, K), 2.0_Real);
```
Don't know how efficient Kokkos::pow is, but often for low integral powers (square, cube), actually doing the multiply is faster than using the pow function.

katsmith133 replied:
Sounds good, I will change it to the multiply

vanroekel commented on components/omega/doc/devGuide/VerticalMixingCoeff.md:
```diff
@@ -0,0 +1,96 @@
+
+Omega includes a `VertMix` class that provides functions that compute `VertDiff` and `VertVisc`, the
+vertical diffusivity and viscosity, where both are defined at the top of cell centers. Currently the
+values of `VertDiff` and `VertVisc` are calculated using the combination of three options: (1) a
```
```suggestion
values of `VertDiff` and `VertVisc` are calculated using the linear combination of three options: (1) a
```

vanroekel commented on components/omega/doc/devGuide/VerticalMixingCoeff.md:
```diff
@@ -0,0 +1,96 @@
+vertical diffusivity and viscosity, where both are defined at the top of cell centers. Currently the
+values of `VertDiff` and `VertVisc` are calculated using the combination of three options: (1) a
+constant background mixing value, (2) a convective instability mixing value, and (3) a Richardson
+number dependent shear mixing value based upon Pacanowski and Philander (1981). These options are
```
```suggestion
number dependent shear mixing value from the Pacanowski and Philander (1981) parameterization. These options are
```

vanroekel replied:
also consider linking to the original paper here or the cvmix docs

vanroekel commented on components/omega/doc/devGuide/VerticalMixingCoeff.md:
```diff
@@ -0,0 +1,96 @@
+number dependent shear mixing value based upon Pacanowski and Philander (1981). These options are
+additive. For both the convective and shear mixing values `BruntVaisalaFreq` is needed, which
+is calculated by the `EOS` class. Currently, the `VertMix` class is set up to have two enumerations,
+`PP` and `KPP`, where the first only uses the three options listed above and the second uses the K Profile Parameterization [(KPP; Large et al., 1994)](https://agupubs.onlinelibrary.wiley.com/doi/abs/10.1029/94rg01872). The `KPP` option is not yet functional (it will throw an error if used), but the framework for its development have been added here.
```
this may be out of scope here, but will KPP also be additive?  We likely want cases with KPP in the OSBL, conv instability and background below.  The text suggests you only get all three with PP as the option

katsmith133 replied:
Yes, I was planning for KPP to be additive as well, but wasn't sure if we wanted to keep the KPP stuff in this PR yet, so didn't go into detail. I will edit the text to reflect the fact that all of them are additive.

philipwjones commented on components/omega/src/ocn/VertMix.h:
```diff
@@ -0,0 +1,210 @@
+   bool Enabled = true; ///< Enable convective mixing flag
+
+   // Convective mixing parameters
+   Real ConvDiff =
```
I realize that these (and similar parameters for other options) will be overwritten by config values, but might want to initialize to zero or some unphysical value to cause it to fail or produce bad results if not initialized later.

philipwjones reviewed (changes requested):
This fails on Frontier GPU with a memory access error, so some array/variable may be missing or not scoped?  I have some other comments spread below.

vanroekel commented on components/omega/doc/devGuide/VerticalMixingCoeff.md:
````diff
@@ -0,0 +1,96 @@
+An enumeration listing of the schemes is provided (`KPP` is not yet functional). It needs to be extended every time a `VertMix` type is added. It is used to identify which VertMix method is to be used at run time.
+
+```c++
+enum class VertMixType { PP, KPP };
````
type PP may not be the best name.  It implies you only get PP81 shear mixing.  I'm also unclear why we need an enum type here for PP vs KPP.  Couldn't we have one vertmix class that has methods for PP, KPP, etc..?  Maybe this will become clearer further down.

katsmith133 replied:
Yup, I think yours and Phil's comments make sense and I will get rid of the different classes and just combine them all. This will automatically get rid of the PP naming as well.

vanroekel commented on components/omega/doc/devGuide/VerticalMixingCoeff.md:
```diff
@@ -0,0 +1,96 @@
+This method combines the effects of:
+- Background mixing (constant coefficients)
+- Convective mixing (triggered by static instability)
+- Shear mixing (Pacanowski-Philander scheme)
```
```suggestion
- Shear instability driven mixing (Pacanowski-Philander scheme)
```

vanroekel commented on components/omega/doc/userGuide/EOS.md:
```diff
@@ -19,4 +19,12 @@ Eos:
+
+## Brunt Vaisala Frequency
+
+The `Eos` class also calculates the Brunt Vaisala Frequency, which is used by the `VertMix` class to determine water column stability for both the convective adjustment and shear-induced mixing. When using the `Linear` EOS option, the `BruntVaisalaFreq` is calculated using with linear coefficients. When using the `Teos-10` EOS option, the `BruntVaisalaFreq` is calculated with non-linear coefficie ...
```
```suggestion
The `Eos` class also calculates the Brunt Vaisala Frequency, which is used by the `VertMix` class to determine water column stability for both the convective adjustment and shear-instability driven mixing. When using the `Linear` EOS option, the `BruntVaisalaFreq` is calculated using with linear coefficients. When using the `Teos-10` EOS option, the `BruntVaisalaFreq` is calculated with non-linear coefficients according to the TOES-10. For both options, `SpecVol` must be calculated prior to calculating `BruntVaisalaFreq`, as it is an input to the `BruntVaisalaFreq` calculation.
```

vanroekel commented on components/omega/doc/userGuide/VerticalMixingCoeff.md:
```diff
@@ -0,0 +1,67 @@
+in the ocean model. Currently, there is one primary operational vertical mixing option, `PP`,
+which can superimpose three different mixing processes within the water column: (1) a constant
+background mixing value, (2) a convective instability mixing value, and (3) a Richardson number
+dependent shear mixing value based upon Pacanowski and Philander (1981). These are describe a bit
```
```suggestion
dependent shear instability driven mixing value based upon Pacanowski and Philander (1981). These are describe a bit
```

vanroekel commented on components/omega/doc/userGuide/VerticalMixingCoeff.md:
```diff
@@ -0,0 +1,67 @@
+    Diffusivity: 1.0     # Convective mixing coefficient (m²/s)
+    TriggerBVF: -1e-4    # Brunt-Vaisala frequency threshold
+  Shear:
+    Enable: true         # Enables the shear-induced mixing option
```
```suggestion
    Enable: true         # Enables the shear instability driven mixing option
```

vanroekel commented on components/omega/doc/userGuide/VerticalMixingCoeff.md:
```diff
@@ -0,0 +1,67 @@
+
+### 2. Convective Mixing
+
+Enhanced convective adjustment mixing that occurs in statically unstable regions of the water column to parameterize convection and homogenize properties. In OMEGA this is mixing is defaulted to occur when the Brunt Vaisala Frequency is less than 0.0 (unstable), and is off when equal to (neutral) or greater than (stable) 0.0.
```
```suggestion
Enhanced convective adjustment mixing that occurs in statically unstable regions of the water column to parameterize convection and homogenize properties. In Omega this is mixing is defaulted to occur when the Brunt Vaisala Frequency is less than 0.0 (unstable), and is off when equal to (neutral) or greater than (stable) 0.0.
```

vanroekel commented on components/omega/doc/userGuide/VerticalMixingCoeff.md:
```diff
@@ -0,0 +1,67 @@
+\kappa_{b} + \kappa_{conv} \quad \text{ if } N^2 < N^2_{crit}\\
+\kappa_{b} \quad \text{ if } N^2 \geq N^2_{crit}
+\end{cases}
+$$
```
you may want to comment that N2 < N2_crit is different than currently in MPAS/CVMix (it's <=)

vanroekel commented on components/omega/src/ocn/Eos.h:
```diff
@@ -261,11 +261,273 @@ class LinearEos {
+      }
+   }
+
+   KOKKOS_FUNCTION Real calcGrav(Real Lat, Real Z) const {
```
do we really want to calculate latitude/depth varying gravity in here when the rest of the code uses a fixed value?

katsmith133 replied:
@vanroekel This is how TEOS-10 does it. If we are ok with the error (compared to the TEOS-10 value) that will incur, than I am ok with changing this to a constant value.

vanroekel replied:
My guess is this variation is not big enough to warrant explicit calculation, especially since it's between two bounding layers.  From what I can tell MOM6 still uses fixed gravity for N2.  It seems inconsistent to use variable gravity here and fixed gravity everywhere else

katsmith133 replied:
Ok, sounds good, I will change it to a constant value.

katsmith133 replied:
The answer is ~0.25% different if Gravity is constant

vanroekel commented on components/omega/src/ocn/Eos.h:
```diff
@@ -261,11 +261,273 @@ class LinearEos {
+   //   and outputs the Brunt-Vaisala frequency.
+   KOKKOS_FUNCTION void operator()(Array2DReal BruntVaisalaFreq, I4 ICell,
+                                   const Array2DReal &SpecVol) const {
+      Real Gravity = 9.80616; // gravitational acceleration
```
should use shared constant here

katsmith133 replied:
@vanroekel I agree, but that PR has not gone through yet

katsmith133 committed:
add review suggestions/edits

katsmith133 commented:
> This fails on Frontier GPU with a memory access error, so some array/variable may be missing or not scoped? I have some other comments spread below.

@philipwjones did the error provide any more information than that? At the moment my Frontier build is stopping at the cmake step (it keeps hanging at the `Cpptrace auto config: Using cxxabi for demangling`?), so I can't test to see what this issue is.

philipwjones commented:
@katsmith133 No additional info and I didn't have a chance to narrow it further, but I forgot to mention that it failed both EOS and VertMix unit tests, so it could point to something in the EOS additions. And I think this was with the AMD compiler - I can look into it further next week if no one gets to it first.

katsmith133 commented:
> @katsmith133 No additional info and I didn't have a chance to narrow it further, but I forgot to mention that it failed both EOS and VertMix unit tests, so it could point to something in the EOS additions. And I think this was with the AMD compiler - I can look into it further next week if no one gets to it first.

Thanks @philipwjones! I am in a Deep Learning course this week, and probably will not get to it. So if you have time, that would be appreciated! If not, I can pick it back up next week.

mwarusz commented on components/omega/test/ocn/VertMixTest.cpp:
```diff
@@ -0,0 +1,658 @@
+
+   parallelFor(
+       "populateArrays", {Mesh->NCellsAll, VCoord->NVertLayers},
+       KOKKOS_LAMBDA(I4 ICell, I4 K) { VCoord->ZMid(ICell, K) = -K; });
```
```suggestion
   OMEGA_SCOPE(ZMid, VCoord->ZMid);
   parallelFor(
       "populateArrays", {Mesh->NCellsAll, VCoord->NVertLayers},
       KOKKOS_LAMBDA(I4 ICell, I4 K) { ZMid(ICell, K) = -K; });
```

mwarusz commented on components/omega/test/ocn/VertMixTest.cpp:
```diff
@@ -0,0 +1,658 @@
+   deepCopy(TestVertMix->VertDiff, 0.0);
+   deepCopy(TestVertMix->VertVisc, 0.0);
+
+   parallelFor(
+       "populateArrays", {Mesh->NCellsAll, VCoord->NVertLayers},
+       KOKKOS_LAMBDA(I4 ICell, I4 K) {
+          VCoord->ZMid(ICell, K)    = -K;
+          Mesh->NEdgesOnCell(ICell) = 5;
+          Mesh->AreaCell(ICell)     = 3.6e10_Real;
+       });
```
```suggestion
   const auto &ZMid = VCoord->ZMid;
   const auto &NEdgesOnCell = Mesh->NEdgesOnCell;
   const auto &AreaCell = Mesh->AreaCell;
   parallelFor(
       "populateArrays", {Mesh->NCellsAll, VCoord->NVertLayers},
       KOKKOS_LAMBDA(I4 ICell, I4 K) {
          ZMid(ICell, K)    = -K;
          NEdgesOnCell(ICell) = 5;
          AreaCell(ICell)     = 3.6e10_Real;
       });
```

philipwjones replied:
Thanks @mwarusz  I can make those changes.

mwarusz reviewed (commented):
@katsmith133 @philipwjones 

I had a chance to look into the test failures of this PR on Frontier. There are issues with accessing host pointers on the device in `EosTest.cpp` and `VertMixTest.cpp`. My suggestion shows how to fix one loop, but similar changes need to be applied for all accesses to `VCoord->ZMid` and various `Mesh->X` arrays in these two tests.

cbegeman commented on components/omega/src/ocn/VertMix.cpp:
```diff
@@ -0,0 +1,270 @@
+   auto VertViscField =
+       Field::create(VertViscFldName, // Field name
+                     "Vertical viscosity at center and"
+                     " top of cell",                   // Long Name
```
```suggestion
                     "Vertical viscosity at center of cell and"
                     " top of layer",                   // Long Name
```

philipwjones committed:
fix frontier GPU issue and add checks to EOS test driver

  - added missing OMEGA_SCOPE statements
  - added some checks/tests to EOS driver
  - modified some error handling

philipwjones commented:
Just pushed some changes to fix the frontier issue in EOS test driver (Vmix test driver mods will come soon). The commit also adds some additional EOS checks and cleans up some error handling, so folks should pull these changes if working on this PR

mwarusz commented on components/omega/src/ocn/VertMix.h:
```diff
@@ -0,0 +1,183 @@
+         } else {
+            if (BruntVaisalaFreq(ICell, K) < ConvTriggerBVF) {
+               VertDiff(ICell, K) = VertDiff(ICell, K) + ConvDiff;
+               VertVisc(ICell, K) = VertVisc(ICell, K) + ConvDiff;
```
```suggestion
               VertDiff(ICell, K) += ConvDiff;
               VertVisc(ICell, K) += ConvDiff;
```

mwarusz commented on components/omega/src/ocn/VertMix.h:
```diff
@@ -0,0 +1,183 @@
+            VertDiff(ICell, K) = 0.0_Real;
+         } else {
+            Real ShearViscVal = 0.0;
+            Real InvAreaCell  = 1.0 / AreaCell(ICell);
```
```suggestion
            Real InvAreaCell  = 1.0_Real / AreaCell(ICell);
```

mwarusz commented on components/omega/src/ocn/VertMix.h:
```diff
@@ -0,0 +1,183 @@
+            VertVisc(ICell, K) = VertVisc(ICell, K) + ShearViscVal;
+            VertDiff(ICell, K) =
+                VertDiff(ICell, K) +
+                ShearViscVal / (1.0_Real + ShearAlpha * RichardsonNum);
```
```suggestion
            VertVisc(ICell, K) += ShearViscVal;
            VertDiff(ICell, K) += ShearViscVal / (1.0_Real + ShearAlpha * RichardsonNum);
```

mwarusz replied:
```suggestion
            VertVisc(ICell, K) += ShearViscVal;
            VertDiff(ICell, K) += ShearViscVal / (1.0_Real + ShearAlpha * RichardsonNum);
```

mwarusz commented on components/omega/src/ocn/Eos.h:
```diff
@@ -197,25 +197,25 @@ class Teos10Eos {
-   KOKKOS_FUNCTION Real calcDelta(const Array2DReal &SpecVolPCoeffs, const I4 K,
-                                  const Real P) const {
+   KOKKOS_FUNCTION Real calcDelta(const Real (&SpecVolPCoeffs)[6 * VecLength],
+                                  const I4 K, const Real P) const {
```
Rename `K` to `KVec` in this function since it is no longer an absolute index.

mwarusz commented on components/omega/src/ocn/Eos.h:
```diff
@@ -44,41 +42,43 @@ class Teos10Eos {
    /// Calculate pressure polynomial coefficients for TEOS-10
-   KOKKOS_FUNCTION void calcPCoeffs(Array2DReal SpecVolPCoeffs, const I4 K,
-                                    const Real Ct, const Real Sa) const {
+   KOKKOS_FUNCTION void calcPCoeffs(Real (&SpecVolPCoeffs)[6 * VecLength],
+                                    const I4 K, const Real Ct,
+                                    const Real Sa) const {
```
Rename `K` to `KVec` in this function since it is no longer an absolute index.

mwarusz commented on components/omega/src/ocn/VertMix.cpp:
```diff
@@ -0,0 +1,270 @@
+                 )
+    : ComputeVertMixConv(VCoord), ComputeVertMixShear(Mesh, VCoord) {
+   VertDiff = Array2DReal("VertDiff", Mesh->NCellsAll, VCoord->NVertLayers + 1);
+   VertVisc = Array2DReal("VertVisc", Mesh->NCellsAll, VCoord->NVertLayers + 1);
```
Why is the vertical extent `VCoord->NVertLayers + 1` ? This is inconsistent with how the fields associated with these arrays are defined later.

mwarusz commented on components/omega/src/ocn/VertMix.h:
```diff
@@ -0,0 +1,183 @@
+                                   I4 ICell,
+                                   const Array2DReal &BruntVaisalaFreq) const {
+      for (int K = 1; K < NVertLayers; ++K) {
+         if (K == 0) {
```
Why is this a serial loop ? Also `K` is never zero, so the first branch of the `if` statement is never used.

mwarusz commented on components/omega/src/ocn/Eos.h:
```diff
@@ -261,11 +261,258 @@ class LinearEos {
+
+      Real Gravity = 9.80616;
+      Real Db2Pa   = 1.0e4;
+      for (int K = 0; K < NVertLayers; ++K) {
```
Why is this computation done serially ?

mwarusz commented on components/omega/src/ocn/Eos.h:
```diff
@@ -261,11 +261,258 @@ class LinearEos {
+   KOKKOS_FUNCTION void operator()(Array2DReal BruntVaisalaFreq, I4 ICell,
+                                   const Array2DReal &SpecVol) const {
+      Real Gravity = 9.80616; // gravitational acceleration
+      for (int K = 0; K < NVertLayers; ++K) {
```
Why is this computation done serially ?

mwarusz commented on components/omega/src/ocn/VertMix.cpp:
```diff
@@ -0,0 +1,270 @@
+             LocVertVisc(ICell, 0) = 0.0_Real;
+          });
+   }
+   Kokkos::fence();
```
```suggestion
```
Shouldn't be necessary.

mwarusz commented on components/omega/src/ocn/VertMix.h:
```diff
@@ -0,0 +1,183 @@
+                                   const Array2DReal &TangentialVelocity,
+                                   const Array2DReal &BruntVaisalaFreq) const {
+
+      for (int K = 0; K < NVertLayers; ++K) {
```
Why is this a serial loop ?

mwarusz commented on components/omega/src/ocn/VertMix.cpp:
```diff
@@ -0,0 +1,270 @@
+   Error Err; // error code
+
+   /// Retrieve default VertMix
+   VertMix *vertmix = VertMix::getInstance();
```
```suggestion
   VertMix *DefVertMix = VertMix::getInstance();
```
Variable names in Omega are capitalized.

mwarusz commented on components/omega/src/ocn/VertMix.h:
```diff
@@ -0,0 +1,183 @@
+   VertMix &operator=(VertMix &&)      = delete;
+
+   I4 NCellsAll;   ///< Number of horizontal cells
+   I4 NEdgesAll;   ///< Number of horizontal edges
```
The `NEdgesAll` member is never used. In fact, it is never even initialized. The same is true for `NVertLayers`, but that variable may be needed if this class is changed to do parallel computations in the vertical.

mwarusz reviewed (commented):
I have a few more comments and suggestions. There is a couple of serial vertical loops in this PR that I think could be easily done in parallel.

philipwjones committed:
fix frontier gpu issues and cleanup Vmix test driver

  - added needed OMEGA_SCOPE calls for some vars
  - update error handling
  - expanded error checks to full arrays
  - cleaned up some variable names

philipwjones commented:
Ok, I pushed a new version of the vert mix test driver that fixes the Frontier issues, expands some error checks to full arrays, and cleans up a few things. This now passes ctests on Frontier.

katsmith133 committed:
add more review suggestions/edits

katsmith133 commented:
@philipwjones Thanks for the help in fixing the Frontier issues!

@mwarusz, @cbegeman, and @vanroekel I believe I have addressed all of your comments. Please let me know if not. Thanks!

philipwjones reviewed (approved):
I'm fine with this now.

vanroekel commented on components/omega/doc/devGuide/VerticalMixingCoeff.md:
```diff
@@ -0,0 +1,86 @@
+2. Convective Mixing:
+   - `EnableConvMix`: Flag to enable/disable convective mixing (Default: True)
+   - `ConvDiff`: Convective mixing coefficient (m²/s; Default: 1.0)
+   - `ConvTriggerBVF`: Trigger threshold for convective mixing (Default: < 0.0)
```
should we be more specific on the default here?  Maybe specify something like -1e-8 or some small negative?

katsmith133 replied:
Oh, yea, it should be just `0.0` not `< 0.0`. I'll change that.

vanroekel commented on components/omega/doc/userGuide/EOS.md:
```diff
@@ -4,7 +4,7 @@
 The equation of state (EOS) for the ocean describes the relationship between specific volume of seawater (in $\textrm{m}^3/\textrm{kg}$; the reciprocal of density) and temperature (in $^{\circ}\textrm{C}$), salinity (in $\textrm{g/kg}$), and pressure (in $\textrm{dbar}$). Through the hydrostatic balance (which relates density/specific volume gradients to pressure gradients), the equation of state ...
 
-Two choices of EOS are provided by Omega: a linear EOS and a TEOS-10 EOS. The linear EOS simplifies the relationship by excluding the influence of pressure and using constant expansion/contraction coefficients, making the specific volume a simple linear function of temperature and salinity. However, this option is only recommended for simpler idealized test cases as its accuracy is not sufficient ...
+Two choices of EOS are provided by Omega: a linear EOS and a TEOS-10 EOS. The linear EOS simplifies the relationship by excluding the influence of pressure and using constant expansion/contraction coefficients, making the specific volume a simple linear function of temperature and salinity. However, this option is only recommended for simpler idealized test cases as its accuracy is not sufficient ...
```
```suggestion
Two choices of EOS are provided by Omega: a linear EOS and a TEOS-10 EOS. The linear EOS simplifies the relationship by excluding the influence of pressure and using constant expansion/contraction coefficients, making the specific volume a simple linear function of temperature and salinity. However, this option should only be used for simpler idealized test cases as it is not appropriate for realistic ocean simulations. The TEOS-10 EOS is a 75-term polynomial expression from [Roquet et al. 2015](https://www.sciencedirect.com/science/article/pii/S1463500315000566) that approximates the [Thermodynamic Equation of Seawater 2010](https://www.teos-10.org/pubs/TEOS-10_Manual.pdf), but in a less complex and more computationally efficient manner, and is the preferred EOS for real ocean simulations in Omega.
```

vanroekel replied:
suggesting stronger language to not use linear for realistic cases

vanroekel commented on components/omega/doc/userGuide/VerticalMixingCoeff.md:
```diff
@@ -0,0 +1,68 @@
+  Convective:
+    Enable: true         # Enables the convective-induced mixing option
+    Diffusivity: 1.0     # Convective mixing coefficient (m²/s)
+    TriggerBVF: -1e-4    # Brunt-Vaisala frequency threshold
```
this seems to large.

katsmith133 replied:
Huh... not sure what happened, but it should be `TriggerBVF: 0.0`. I'll fix that.

vanroekel commented on components/omega/src/ocn/Eos.h:
```diff
@@ -261,11 +262,266 @@ class LinearEos {
+            Real PInt =
+                0.5_Real * (Pressure(ICell, K) + Pressure(ICell, K - 1));
+            Real SpInt = 0.5_Real * (SpecVol(ICell, K) + SpecVol(ICell, K - 1));
+            Real AlphaInt = calcAlpha(SaInt, CtInt, PInt, SpInt);
```
we may want to put alpha and beta in member arrays as KPP will need these too and then we won't have to recalculate (I think).

katsmith133 replied:
I'll keep this as a note to change when we implement KPP. Thanks for pointing it out!

vanroekel commented on components/omega/src/ocn/HorzMesh.cpp:
```diff
@@ -643,6 +643,7 @@ void HorzMesh::copyToDevice() {
    YCell             = createDeviceMirrorCopy(YCellH);
    XEdge             = createDeviceMirrorCopy(XEdgeH);
    YEdge             = createDeviceMirrorCopy(YEdgeH);
+   LatCell           = createDeviceMirrorCopy(LatCellH);
```
is this still needed when not computing the spatially variable gravity?

vanroekel commented on components/omega/src/ocn/HorzMesh.h:
```diff
@@ -163,6 +163,8 @@ class HorzMesh {
    HostArray1DReal ZCellH;   ///< Z Coordinates of cell centers (m)
    HostArray1DReal LonCellH; ///< Longitude location of cell centers (radians)
+
+   Array1DReal LatCell;      ///< Latitude location of cell centers (radians)
```
same comment.

vanroekel commented:
This is looking great @katsmith133! I had just a few further comments that I've added.

katsmith133 committed:
add more review suggestions/edits

mwarusz commented on components/omega/doc/devGuide/VerticalMixingCoeff.md:
````diff
@@ -0,0 +1,86 @@
+
+```c++
+// Initialize VertMix
+VertMix* vertMix = VertMix::getInstance("Default", mesh, nLevels);
````
```suggestion
VertMix* VMix = VertMix::getInstance();
```

mwarusz commented on components/omega/doc/devGuide/VerticalMixingCoeff.md:
```diff
@@ -0,0 +1,86 @@
+
+// Compute mixing coefficients
+vertMix->computeVertMix(VertDiff, VertVisc, NormalVelocity,
+                       TangentialVelocity, BruntVaisalaFreq);
```
```suggestion
VMix->computeVertMix(NormalVelocity,
                       TangentialVelocity, BruntVaisalaFreq);
```

mwarusz commented on components/omega/doc/devGuide/VerticalMixingCoeff.md:
```diff
@@ -0,0 +1,86 @@
+                   Array2DReal VertVisc,
+                   const Array2DReal &NormalVelocity,
+                   const Array2DReal &TangentialVelocity,
+                   const Array2DReal &BruntVaisalaFreq);
```
```suggestion
void computeVertMix(
                   const Array2DReal &NormalVelocity,
                   const Array2DReal &TangentialVelocity,
                   const Array2DReal &BruntVaisalaFreq);
```

mwarusz reviewed (approved):
Thanks @katsmith133 for addressing my comments. I ran the ctests on pm-cpu, pm-gpu and frontier-gpu and they passed. Other than some small issues in the dev docs, this looks fine so I am approving.

katsmith133 committed:
updates user and dev guides

cbegeman reviewed (approved):
@katsmith133 Thanks for all your work on this! Just noting that the scope of this PR does not include applying an update to the state at runtime. Thus, we will run Polaris tests after the vertical mixing update is applied during time stepping.

vanroekel reviewed (approved):
Thanks for the great work @katsmith133! I have no further comments

alicebarthel commented on components/omega/doc/userGuide/EOS.md:
```diff
@@ -19,4 +19,12 @@ Eos:
+
+## Displaced Specific Volume
+
+The `Eos` class calculates the density of a parcel of fluid that is adiabatically displaced by a relative `k` levels, capturing the effects of pressure/depth changes. This is primarily used to calculate quantities for determining the water column stability (i.e. the stratification) and the vertical mixing coefficients (viscosity and diffusivity). Note: when using the `Linear` EOS option, `SpecVol ...
```
K counted positive downward. 
```suggestion
The `Eos` class calculates the density of a parcel of fluid that is adiabatically displaced by a relative `k` levels, k counted positive downward, capturing the effects of pressure/depth changes. This is primarily used to calculate quantities for determining the water column stability (i.e. the stratification) and the vertical mixing coefficients (viscosity and diffusivity). Note: when using the `Linear` EOS option, `SpecVolDisplaced` will be the same as `SpecVol` since the specific volume calculation is independent of pressure/depth.
```

alicebarthel commented on components/omega/doc/devGuide/EOS.md:
```diff
@@ -2,12 +2,19 @@
+If `SpecVolDisplaced` is calculated with the linear EOS option, it will be equal to `SpecVol` as there
 is no pressure/depth dependence for the linear EOS. `SpecVolDisplaced` computes specific volume
-adiabatically displaced to `K + KDisp`.
+adiabatically displaced to `K + KDisp`. Note: `SpecVol` must be calculated before `BruntVaisalaFreq`, as
```
```suggestion
adiabatically displaced to `K + KDisp` (K counted positive downward, ie K+1 is one layer below K). Note: `SpecVol` must be calculated before `BruntVaisalaFreq`, as
```

alicebarthel commented on components/omega/doc/userGuide/VerticalMixingCoeff.md:
```diff
@@ -0,0 +1,68 @@
+
+$$
+Ri = \frac{N^2}{\left|\frac{\partial \mathbf{U}}{\partial z}\right|^2}\,,
+$$
```
```suggestion
$$
where $\nu_o$, $\alpha$, $n$, $\nu_b$, $\kappa_b$ are constant parameters set in the yaml file. $N^2$ is calculated by the EOS based on the ocean state and $\mathbf{U}$ is the magnitude of the horizontal velocity. $N^2$, $\partial \mathbf{U}}{\partial z}\right|^2$ and $Ri$ of K are all defined at the top interface of layer K.  $N^2$, $\nu_{shear}$ and $\kappa_{shear}$ are set to zero for the surface layer.
```

alicebarthel replied:
@katsmith133 here my suggested $\nu_{shear}$ and $\kappa_{shear}$ are not yet defined. You will want to update that depending on what we decide regarding the notations of $\nu_{shear}$  vs $\nu$, but hopefully you know what I am trying to say in my suggestion above.

alicebarthel commented on components/omega/src/ocn/Eos.cpp:
```diff
@@ -176,19 +184,60 @@ void Eos::computeSpecVolDisp(const Array2DReal &ConservTemp,
 
    /// Create fields for state variables
-   int NDims = 2;
+   const Real FillValue = -9.99e30;
```
out of curiosity, is it something that we've agreed on for a Fill Value?

alicebarthel commented on components/omega/src/ocn/Eos.h:
```diff
@@ -44,41 +42,43 @@ class Teos10Eos {
                                    I4 KDisp) const {
 
-      OMEGA_SCOPE(LocSpecVolPCoeffs, SpecVolPCoeffs);
+      Real SpecVolPCoeffs[6 * VecLength];
```
@katsmith133 Can you tell me more about why the SpecVolPCoeffs were moved into the operator rather than as a straightforward class data member? I'm sure that there was a reason, just to build my understanding. Do the coefficients still remain public for other member functions to reuse them?

alicebarthel replied:
My sense is that the main goal is to save on memory by having a [6 * VecLength] array instead of a full K array (local vs global vertical index). If that means that we cannot reuse the coeffs, it becomes a performance vs memory issue: is it worth saving a K-array and compute the coefficients once (instead of 3?). We may want to discuss or test this. Or make a note for later optimization if we need to move forward with this.

alicebarthel replied:
From discussion with Kat: this was done in the process of resolving an issue on GPU. It may not be necessary but is optimal for this version of the code. We will revisit (and likely revert to a data variable) when we test optimizing the performance by reusing PCoeffs.

alicebarthel commented on components/omega/src/ocn/Eos.h:
```diff
@@ -261,11 +262,266 @@ class LinearEos {
+                                   const Array2DReal &Pressure,
+                                   const Array2DReal &SpecVol) const {
+      Real Gravity    = 9.80616;
+      Real Db2Pa      = 1.0e4;
```
Not a problem in itself but some functions assume P in dbars and other assume it in Pa with a conversion to Db. I think it is fine as it is here but let's make it consistent across the Eos and explicitly state it in the documentation.

alicebarthel replied:
This is correct because Db2Pa is only applied to DP. It is never used to modify the pressure field itself (calcAlpha is applied to PInt which is still in dbars.

alicebarthel commented on components/omega/src/ocn/Eos.h:
```diff
@@ -261,11 +262,266 @@ class LinearEos {
+   KOKKOS_FUNCTION Real calcAlpha(Real Sa, Real Ct, Real P, Real Sp) const {
+      constexpr Real Factor = 0.0248826675584615;
+      constexpr Real Offset = 5.971840214030754e-1;
+      constexpr Real Pa2Db  = 1.0e-4;
```
Sorry to harp on this. Is `Pa2Db` your interpretation of the 1/Pu? Or a conversion factor you added @katsmith133? It might be correct but the teos-10 documentation still states that they expect P in dbars. They might be wrong but I'd want to be sure we are tracking the right coefficients and units throughout...

alicebarthel replied:
Following up on our conversation, I still believe that calling this `Pa2Db` is confusing and unnecessarily so. I couldn't find the code you were pointing to but the coefficients you are using are all in line with say, [alpha code](https://github.com/TEOS-10/python-gsw/blob/7d6ebe8114c5d8b4a64268d36100a70e226afaf6/gsw/gibbs/density_enthalpy_75.py#L158-L159), which apply to P in dbars. So the field P in dbars is multiplied by 1.0e-4, but it's not to convert Pa in Dbars.

alicebarthel commented on components/omega/src/ocn/Eos.h:
```diff
@@ -197,25 +197,26 @@ class Teos10Eos {
+                                  const I4 KVec, const Real P) const {
 
       constexpr Real Pu = 1e4;
       Real Pp           = P / Pu;
```
This assumes P in dbars (according to the code headers).  If we assume we pass it P in Pa then let's update this to be consistent. Or triple check the units and coefficients.

alicebarthel commented on components/omega/src/ocn/VertMix.h:
```diff
@@ -0,0 +1,190 @@
+                                          ShearExponent);
+            VertVisc(ICell, K) += ShearViscVal;
+            VertDiff(ICell, K) +=
+                ShearViscVal / (1.0_Real + ShearAlpha * RichardsonNum);
```
There is an inconsistency here between the documentation and the implementation. In the documentation, `VertDiff` is based on `VertVisc` which includes the background viscosity (at least the way it is written). In the implementation `ShearViscVal` does not include the background viscosity and is passed to calculate `VertDiff`. Can we adjust for consistency? The paper is in line with the documentation. I didn't find the missing `BackVisc / (1.0_Real + ShearAlpha * RichardsonNum)` added to `VertDiff` anywhere.

alicebarthel replied:
@katsmith133 To answer our question,`BackVisc` **is included** in the cvmix version of PP Kappa ([see code here](https://github.com/CVMix/CVMix-src/blob/9187de96a40ca55ec7b43524a2bf0ace10bcae67/src/shared/cvmix_shear.F90#L400-L401)), as the paper describes. 
Then we remove the background values [here](https://github.com/E3SM-Project/E3SM/blob/8cbc1d3a556da0b21522978aaba84aa5e39c1d91/components/mpas-ocean/src/shared/mpas_ocn_vmix_cvmix.F#L452-L453) in mpas-ocean to avoid double counting the background diffusivity. But $\kappa_{shear}$ (aka `cvmix_variables % Tdiff_iface(k)`) still includes some influence of `PP_nu_b` in our present implementation. 
So if we want to implement the same way in Omega1 (cvmix as implemented in mpas-ocean), just add the extra bit as I suggested above. If people feel they would rather have $\kappa_{shear}$ completely independent of $\nu_b$, we can keep the code in its current version but should explicitly document that decision.

vanroekel replied:
Maybe we could just skip the PP81 implementation?  By default we use the interior mixing shear instability driven scheme of LMD94.  I doubt it would be that different from PP81 and doesn't have this backVisc issue.

alicebarthel commented on components/omega/src/ocn/Eos.h:
```diff
@@ -261,11 +262,266 @@ class LinearEos {
+            Real DP       = Pressure(ICell, K) - Pressure(ICell, K - 1);
+
+            BruntVaisalaFreq(ICell, K) = Gravity * Gravity *
+                                         ((1.0_Real / SpInt) / (Db2Pa * DP)) *
```
It's a minor thing but wouldn't it make sense to keep the straightforward expression that match the [teos-10 doc](https://www.teos-10.org/pubs/gsw/html/gsw_Nsquared.html)?
```suggestion
                                         (BetaInt * DSa - AlphaInt * DCt)/ (SpInt * Db2Pa * DP)
```

alicebarthel commented on components/omega/doc/userGuide/EOS.md:
```diff
@@ -19,4 +19,12 @@ Eos:
+
+## Brunt Vaisala Frequency
+
+The `Eos` class also calculates the Brunt Vaisala Frequency, which is used by the `VertMix` class to determine water column stability for both the convective adjustment and shear-instability driven mixing. When using the `Linear` EOS option, the `BruntVaisalaFreq` is calculated using linear coefficients. When using the `Teos-10` EOS option, the `BruntVaisalaFreq` is calculated with non-linear coe ...
```
OK this is one of my pet peeves but we store the Brunt Vaisala Frequency **Squared** in our `BruntVaisalaFreq` variable. Can we rename it `BruntVaisalaFreqS` by any chance? It bothers me to have a "frequency" variable in s-2.

alicebarthel commented on components/omega/src/ocn/Eos.h:
```diff
@@ -261,11 +262,266 @@ class LinearEos {
+            Real DCt      = ConservTemp(ICell, K) - ConservTemp(ICell, K - 1);
+            Real DP       = Pressure(ICell, K) - Pressure(ICell, K - 1);
+
+            BruntVaisalaFreq(ICell, K) = Gravity * Gravity *
```
As mentioned above, I would prefer if this variable was named `BruntVaisalaFreqS` or something to that effect, given that this is a frequency squared, not a frequency. The field metadata is appropriately documented so I'm open to being outvoted on this.

alicebarthel reviewed (changes requested):
Thanks for all your work on this @katsmith133! Most of my comments are suggestions for clarification (in the documentation or the naming). The key point to revise or resolve is whether $\nu_b$ should or should not be included in $\kappa_{shear}$ (see additional comment above on cvmix). We agreed to proceed with the new (smaller) PCoeff array, which will need to be revisited in the performance optimization or the submeso implementation.

katsmith133 committed:
final review suggestions/edits

katsmith133 committed:
updates eos and vert mix  docs

katsmith133 committed:
rebase and adds in hierarchical parallelism to limit vert loops

mark-petersen commented:
Passes all tests on perlmutter and frontier, CPU and GPU, including `Test #36: VERTMIX_TEST`.

<details>
<summary> Test commands: </summary> <p>

```
######### Frontier cpu ############
export CODEDIR=opr
#export CODEDIR=omega-develop
export RUNDIR=test_omega_cpu
mkdir /lustre/orion/cli115/scratch/mpetersen/runs/$RUNDIR
cd !$

cd /ccs/home/mpetersen/repos/E3SM/${CODEDIR}
git submodule update --init --recursive components/omega/external/GSW-C components/omega/external/yaml-cpp externals/YAKL externals/ekat externals/scorpio cime externals/cpptrace
cd /lustre/orion/cli115/scratch/mpetersen/runs/$RUNDIR

module load cmake
rm -rf build
mkdir build
cd build

# compiler options are:
export compiler=craycray
#export compiler=craygnu
#export compiler=crayamd

export PARMETIS_ROOT=/ccs/proj/cli115/software/polaris/frontier/spack/dev_polaris_0_9_0_${compiler}_mpich/var/spack/environments/dev_polaris_0_9_0_${compiler}_mpich/.spack-env/view

module load python
cmake \
   -DOMEGA_CIME_COMPILER=${compiler} \
   -DOMEGA_PARMETIS_ROOT=${PARMETIS_ROOT}\
   -DOMEGA_BUILD_TYPE=Release \
   -DOMEGA_CIME_MACHINE=frontier \
   -DOMEGA_VECTOR_LENGTH=1 \
   -DOMEGA_BUILD_TEST=ON \
   -Wno-dev \
   -S /ccs/home/mpetersen/repos/E3SM/${CODEDIR}/components/omega -B .
# note OMEGA_VECTOR_LENGTH=8 fails MPI tests on CPUs.
./omega_build.sh

# linking:
cd test
ln -isf ~/meshes/omega/O*nc .
cp /ccs/home/mpetersen/repos/E3SM/${CODEDIR}/components/omega/configs/Default.yml omega.yml

# -S is number of GPUs: Count of Specialized Cores per node

salloc -A cli115 -J inter -t 1:00:00 -q debug -N 1 -S 0
cd /lustre/orion/cli115/scratch/mpetersen/runs/$RUNDIR/build
./omega_ctest.sh

######### Frontier gpu #########
export CODEDIR=opr
export RUNDIR=test_omega_gpu
mkdir /lustre/orion/cli115/scratch/mpetersen/runs/$RUNDIR
cd !$

cd /ccs/home/mpetersen/repos/E3SM/${CODEDIR}
git submodule update --init --recursive components/omega/external/GSW-C components/omega/external/yaml-cpp externals/YAKL externals/ekat externals/scorpio cime externals/cpptrace
cd /lustre/orion/cli115/scratch/mpetersen/runs/$RUNDIR

module load cmake
rm -rf build
mkdir build
cd build

# compiler options are:
export compiler=craycray-mphipcc	
#export compiler=craygnu-mphipcc  # 250422 error on cmake, module cmake, git missing
#export compiler=crayamd-mphipcc  

export PARMETIS_ROOT=/ccs/proj/cli115/software/polaris/frontier/spack/dev_polaris_0_9_0_${compiler}_mpich/var/spack/environments/dev_polaris_0_9_0_${compiler}_mpich/.spack-env/view

#module load Core/24.07
#module load cmake/3.27.9 git/2.45.1
module load python
cmake \
   -DOMEGA_CIME_COMPILER=${compiler} \
   -DOMEGA_PARMETIS_ROOT=${PARMETIS_ROOT}\
   -DOMEGA_BUILD_TYPE=Release \
   -DOMEGA_CIME_MACHINE=frontier \
   -DOMEGA_BUILD_TEST=ON \
   -DOMEGA_VECTOR_LENGTH=1 \
   -Wno-dev \
   -Wno-deprecated \
   -S /ccs/home/mpetersen/repos/E3SM/${CODEDIR}/components/omega -B .
./omega_build.sh

# linking:
cd test
ln -isf ~/meshes/omega/O*nc .
cp /ccs/home/mpetersen/repos/E3SM/${CODEDIR}/components/omega/configs/Default.yml omega.yml

salloc -A cli115 -J inter -t 1:00:00 -q debug -N 1 -p batch
# -p is partition name
cd /lustre/orion/cli115/scratch/mpetersen/runs/$RUNDIR/build
./omega_ctest.sh

######### perlmutter CPU
#export CODEDIR=omega-develop
export CODEDIR=opr
export RUNDIR=test_omega_cpu

cd /global/homes/m/mpeterse/repos/E3SM/${CODEDIR}
#git fetch
#git reset --hard origin/develop
git submodule update --init --recursive components/omega/external/GSW-C components/omega/external/yaml-cpp externals/ekat externals/scorpio cime externals/cpptrace
cd components/omega/

module load cmake
mkdir ${PSCRATCH}/runs/$RUNDIR
cd ${PSCRATCH}/runs/$RUNDIR

rm -rf build
mkdir build
cd build

# compiler options are:
export compiler=gnu
#export compiler=nvidia # not working 250421

export PARMETIS_ROOT=/global/cfs/cdirs/e3sm/software/polaris/pm-cpu/spack/dev_polaris_0_9_0_${compiler}_mpich/var/spack/environments/dev_polaris_0_9_0_${compiler}_mpich/.spack-env/view

# nvidia or gnu compiler:
cmake \
   -DOMEGA_CIME_COMPILER=${compiler} \
   -DOMEGA_BUILD_TYPE=Release \
   -DOMEGA_CIME_MACHINE=pm-cpu \
   -DOMEGA_PARMETIS_ROOT=${PARMETIS_ROOT}\
   -DOMEGA_BUILD_TEST=ON \
   -DOMEGA_VECTOR_LENGTH=1 \
   -Wno-dev \
   -S /global/homes/m/mpeterse/repos/E3SM/${CODEDIR}/components/omega -B .
# note OMEGA_VECTOR_LENGTH=8 fails MPI tests on CPUs.
cd ${PSCRATCH}/runs/$RUNDIR/build
./omega_build.sh

# linking:
cd test
ln -isf /global/homes/m/mpeterse/meshes/omega/O*nc .
cp /global/homes/m/mpeterse/repos/E3SM/${CODEDIR}/components/omega/configs/Default.yml omega.yml

# run test:
salloc --nodes 1 --qos interactive --time 01:00:00 --constraint cpu --account=m4572 # or e3sm
cd ${PSCRATCH}/runs/${RUNDIR}/build

./omega_ctest.sh

######### perlmutter GPU
salloc --nodes 4 --qos interactive --time 01:00:00 --constraint gpu --tasks-per-node=2 --gpus-per-task 1 --account=m4572_g # or e3sm_g

# perlmutter has nodes with either 40 or 80 gb of high bandwidth memory, and the system defaults to 40. You can ask for 80 gb nodes with the sbatch flag --constraint="gpu&hbm80gb"

export CODEDIR=opr
#export CODEDIR=omega-develop
export RUNDIR=test_omega_gpu
mkdir ${PSCRATCH}/runs/$RUNDIR
cd !$

rm -rf build
mkdir build
cd build
module load cmake

# compiler options are:
export compiler=gnugpu
#export compiler=nvidiagpu

export PARMETIS_ROOT=/global/cfs/cdirs/e3sm/software/polaris/pm-gpu/spack/dev_polaris_0_9_0_${compiler}_mpich/var/spack/environments/dev_polaris_0_9_0_${compiler}_mpich/.spack-env/view
cmake \
   -DOMEGA_CIME_COMPILER=${compiler} \
   -DOMEGA_BUILD_TYPE=Release \
   -DOMEGA_CIME_MACHINE=pm-gpu \
   -DOMEGA_PARMETIS_ROOT=${PARMETIS_ROOT}\
   -DOMEGA_BUILD_TEST=ON \
   -Wno-dev \
   -DOMEGA_MPI_ON_DEVICE:BOOL=OFF \
   -S /global/homes/m/mpeterse/repos/E3SM/${CODEDIR}/components/omega -B .
# needed for compiler bug: OMEGA_MPI_ON_DEVICE:BOOL=OFF. See https://github.com/E3SM-Project/Omega/issues/214
./omega_build.sh

# linking:
cd test
ln -isf /global/homes/m/mpeterse/meshes/omega/O*nc .
cp /global/homes/m/mpeterse/repos/E3SM/${CODEDIR}/components/omega/configs/Default.yml omega.yml

cd ..
./omega_ctest.sh

```
</details>

mark-petersen commented:
Thank you @katsmith133 and congratulations on the completion of this project.
